csharp/ql/src/experimental/CWE-942/CorsMisconfiguration.ql

```diff
@@ -0,0 +1,58 @@
+/**
+ * @name Credentialed CORS Misconfiguration
+ * @description Allowing any origin while allowing credentials may result in security issues as third party website may be able to
+ *              access private resources.
+ * @kind problem
+ * @problem.severity error
+ * @security-severity 7.5
+ * @precision high
+ * @id cs/web/cors-misconfiguration
+ * @tags security
+ *       external/cwe/cwe-942
+ */
+
+import csharp
+import CorsMisconfigurationLib
+
+/**
+ * Holds if the application allows an origin using "*" origin.
+ */
+private predicate allowAnyOrigin(MethodCall m) {
+  m.getTarget()
+      .hasFullyQualifiedName("Microsoft.AspNetCore.Cors.Infrastructure.CorsPolicyBuilder",
+        "AllowAnyOrigin")
+}
+
+/**
+ * Holds if the application uses a vulnerable CORS policy.
+ */
+private predicate hasDangerousOrigins(MethodCall m) {
+  m.getTarget()
+      .hasFullyQualifiedName("Microsoft.AspNetCore.Cors.Infrastructure.CorsPolicyBuilder",
+        "WithOrigins") and
+  (
+    m.getAnArgument().getValue() = ["null", "*"]
+    or
+    exists(StringLiteral idStr |
+      idStr.getValue().toLowerCase().matches(["null", "*"]) and
+      DataFlow::localExprFlow(idStr, m.getAnArgument())
+    )
+  )
+}
+
+from MethodCall add_policy, MethodCall child
+where
+  (
+    usedPolicy(add_policy) and
+    // Misconfigured origin affects used policy
+    add_policy.getArgument(1) = child.getParent*()
```

Comment:
With this constraint we only consider child *calls* which are *syntactically* included (inlined) in the builder argument.
The following is an example, which would not be caught
```csharp
public class Test
{
  public void ConfigurePolicy(CorsPolicyBuilder policy)
  {
    policy.AllowAnyOrigin().AllowCredentials().AllowAnyHeader().AllowAnyMethod();
  }

  public void ConfigureServices(string[] args)
  {
    var builder = WebApplication.CreateBuilder(args);
    var MyAllowSpecificOrigins = "_myAllowSpecificOrigins";
    builder.Services.AddCors(options =>
    {
      options.AddPolicy(MyAllowSpecificOrigins, policy => ConfigurePolicy(policy));
    });

    var app = builder.Build();

    app.MapGet("/", () => "Hello World!");
    app.UseCors(MyAllowSpecificOrigins);

    app.Run();
  }
}
```
Instead we could use the *calls* relation.
We would need the helper predicate
```
/**
 * Gets the actual callable corresponding to the expression `e`.
 */
Callable getCallableFromExpr(Expr e) {
  exists(Expr dcArg | dcArg = e.(DelegateCreation).getArgument() |
    result = dcArg.(CallableAccess).getTarget() or
    result = dcArg.(AnonymousFunctionExpr)
  )
  or
  result = e
}
```
and then write
```
getCallableFromExpr(add_policy.getArgument(1)).calls*(child.getTarget())
```
The same goes for the other cases, where we need to identify the *child* calls.

Reply:
This is useful thank you! 😮